src/Microsoft.ML.AutoML/AutoMLExperiment/IPerformanceMonitor.cs

```diff
@@ -0,0 +1,114 @@
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+using System.Threading.Tasks;
+using System.Timers;
+using Microsoft.ML.Runtime;
+
+namespace Microsoft.ML.AutoML
+{
+    public interface IPerformanceMonitor : IDisposable
+    {
+        void Start();
+
+        void Stop();
+
+        double? GetPeakMemoryUsageInMegaByte();
+
+        double? GetPeakCpuUsage();
+
+        public event EventHandler<double> CpuUsage;
+
+        public event EventHandler<double> MemoryUsageInMegaByte;
+    }
+
+    internal class DefaultPerformanceMonitor : IPerformanceMonitor
+    {
+        private readonly IChannel _logger;
+        private Timer _timer;
+        private double? _peakCpuUsage;
+        private double? _peakMemoryUsage;
+        private readonly int _checkIntervalInMilliseconds;
+        private TimeSpan _totalCpuProcessorTime;
+
+        public DefaultPerformanceMonitor(IChannel logger, int checkIntervalInMilliseconds)
+        {
+            _logger = logger;
+            _checkIntervalInMilliseconds = checkIntervalInMilliseconds;
+        }
+
+
+        public event EventHandler<double> CpuUsage;
+
+        public event EventHandler<double> MemoryUsageInMegaByte;
+
+
+        public void Dispose()
+        {
+            Stop();
+        }
+
+        public double? GetPeakCpuUsage()
+        {
+            return _peakCpuUsage;
+        }
+
+        public double? GetPeakMemoryUsageInMegaByte()
+        {
+            return _peakMemoryUsage;
+        }
+
+        public void Start()
+        {
+            if (_timer == null)
+            {
+                _timer = new Timer(_checkIntervalInMilliseconds);
+                _totalCpuProcessorTime = Process.GetCurrentProcess().TotalProcessorTime;
+                _timer.Elapsed += OnCheckCpuAndMemoryUsage;
+                _timer.AutoReset = true;
+                _timer.Enabled = true;
+                _logger?.Trace($"{typeof(DefaultPerformanceMonitor)} has been started");
+            }
+        }
+
+        public void Stop()
+        {
+            _timer?.Stop();
+            _timer?.Dispose();
+            _timer = null;
+            _peakCpuUsage = null;
+            _peakMemoryUsage = null;
+        }
+
+        private void OnCheckCpuAndMemoryUsage(object source, ElapsedEventArgs e)
+        {
+            SampleCpuAndMemoryUsage();
+        }
+
+        private void SampleCpuAndMemoryUsage()
+        {
+            // calculate CPU usage in %
+            using (var process = Process.GetCurrentProcess())
+            {
+                var currentCpuProcessorTime = Process.GetCurrentProcess().TotalProcessorTime;
+                var elapseCpuProcessorTime = currentCpuProcessorTime - _totalCpuProcessorTime;
+                var cpuUsedMs = elapseCpuProcessorTime.TotalMilliseconds;
+                var cpuUsageInTotal = cpuUsedMs / (Environment.ProcessorCount * _checkIntervalInMilliseconds);
```

Comment:
>Environment.ProcessorCount * _checkIntervalInMilliseconds [](http://example.com/codeflow?start=51&length=57)

nit: This can be calculated once in the constructor.

Reply:
compiler should be able to optimize that.

Reply:
Are you assuming that? or you have tried it? 

Trying quickly in [csharplab](https://sharplab.io/#v2:C4LghgzgtgPgAgJgAwFgBQiCM71wMwAEiBAKgKYTAHoDe6BDBADgE4CWAbmMGQWwHZUA+gDMA3PUatO3XgGMA9v0p9BBAHIBXKAQC8BZBLSNqxxvlIVgACgCUkhnTMmGovQUxIjJgL4PTJhZwACwEAArsggBqYAA2mmR2/k4ujAD0aapUHO5uAFQEAKL8HGwsSlBkggB0EQpyFBAKLADCCpqC3qlZBDn6+RraXalwmACc1hy2wwR+zgHmCAjJ/ib8ZADulpR2tZHAMfGJ0/5zcyb+QA=) I am seeing it is not optimized. note I didn't use `Environment.ProcessorCount` because it is just 1 and will be optimized at that time. so, I used other constant. 

Anyway, it is a minor point, and you can ignore it if you are ok with the current code.

Reply:
Yeah indeed it's not, turns out I'm too optimistic on the capability of how much optimization compiler can be.

Thanks for introducing csharplab, that's such a wonderful tool